### PR TITLE
lazier, less-jazzy linalg internals

### PR DESCRIPTION
**Pull request by Sacha0**
This pull request completes the first two major blocks of work towards JuliaLang/LinearAlgebra.jl#57. Specifically,

1. The first commit introduces `Adjoint` and `Transpose` types, basic functionality for those types, and an extensive test suite for that functionality.

2. The remaining twenty-one commits then strip all `A[ct]_(mul|ldiv|rdiv)_B[ct][!]` definitions from `Base`. More precisely, these commits transform each such definition to a corresponding `*`/`/`/`\`/`mul!`/`ldiv!`/`rdiv!` definition over `Adjoint`/`Transpose`, and then place the original signature in `base/deprecated.jl` as a shim that calls the transformed definition.

    (In a vain attempt to minimize erosion of my sanity in this process, I made the preceding transformations in the most safe and mechanically simple manner I could: Each method like `A_ldiv_Bc!(A::TypeA, B::TypeB) = $METHBODY` tranforms in place to `ldiv!(A::TypeA, adjB::Adjoint{<:Any,<:TypeB}) = (B = adjB.parent; $METHBODY)` with shim `A_ldiv_Bc!(A::TypeA, B::TypeB) = ldiv!(A, Adjoint(B))` placed in base/deprecated.jl. I also de-meta-fied definitions where that made life simpler.)

All intermediate commits pass the linear algebra and sparse test suites locally.

Collapsing all e.g. `A[ct]_mul_B[ct]` methods into a single e.g. `*` function yields the mother of all ambiguity resolution nightmares. ~So while I fixed all ambiguities necessary to get each commit to pass tests, I expect CI's general method ambiguity filter to detonate on contact.~ Ambiguities resolved.

(Updated.) Next steps, whether in this or subsequent pull requests:

3. ~Strip all `A[ct]_(mul|ldiv|rdiv)_B[ct][!]` definitions from `stdlib/`.~ done here
4. ~Strip all explicit `A[ct]_(mul|ldiv|rdiv)_B[ct][!]` calls from `base/`.~ done here
5. ~Strip all explicit `A[ct]_(mul|ldiv|rdiv)_B[ct][!]` calls from `test/`.~ done here
6. ~Strip all explicit `A[ct]_(mul|ldiv|rdiv)_B[ct][!]` calls from `stdlib/`.~ done here
7. ~Transition linalg to `Adjoint`/`Transpose` externally.~ done in JuliaLang/julia#25083
8. ~Strip all `.'` instances from `base/`.~ done in JuliaLang/julia#25125
9. ~Strip all `.'` instances from `test/`.~ done in JuliaLang/julia#25125
10. ~Strip all `.'` instances from `stdlib/`.~ done in JuliaLang/julia#25125
11. ~Remove special `A[t]_(mul|ldiv|rdiv)_B[t]` lowering.~ done in JuliaLang/julia#25125
12. ~Deprecate `.'`.~ done in JuliaLang/julia#25125
13. ~Rewrite all semantically eager `'` calls in `base/`.~ done in JuliaLang/julia#25148
14. ~Rewrite all semantically eager `'` calls in `test/`.~ done in JuliaLang/julia#25148
15. ~Rewrite all semantically eager `'` calls in `stdlib/`.~ done in JuliaLang/julia#25148
16. ~Change `'` lowering from `adjoint` to `Adjoint`.~ done in JuliaLang/julia#25148
17. ~Remove special `A[c]_(mul|ldiv|rdiv)_B[c]` lowering.~ done in JuliaLang/julia#25148
18. ~Deprecate all the things.~ done in JuliaLang/julia#25217
19. Clean up, docs, news. partial JuliaLang/julia#25217
20. Sleep.

Best!

### CODE REVIEW

**Comment by Sacha0** [comment]
(CI appears happy apart from the global method ambiguity check and stdlib/suitesparse, which I had forgotten.)

**Comment by Sacha0** [comment]
Fixed the one non-ambiguity CI failure. Satisfying the ambiguity check...
```julia
julia> length(Test.detect_ambiguities(Core, Base; imported=true, recursive=true, ambiguous_bottom=false))
387
```
... might take a few days.

**Comment by Sacha0** [comment]
Unsurprisingly, the vast majority of the ambiguities come from transferring the generic fallbacks for `A[ct]_(mul|ldiv|rdiv)_B[ct](a, b)` from base/operators.jl (for example `Ac_mul_B(a, b) = adjoint(a) * b`) to the corresponding consolidation functions; the number of ambiguities falls from 300+ to 99 on dropping the relevant commit (https://github.com/JuliaLang/julia/pull/24969/commits/cbea711e4e9c391b09d18ef80fd293ab7d6d81b0). The resolution signatures for the 99 ambiguities without that commit seem mostly reasonable, e.g. ambiguous pair
```julia
(*(A::AbstractArray{T,2} where T, adjB::Base.LinAlg.Adjoint{#s336,#s335} where #s335<:(Union{Hermitian{T,S}, Hermitian{Complex{T},S}, Symmetric{T,S}} where S where T<:Real) where #s336) in Base.LinAlg at linalg/symmetric.jl:332, *(adjA::Base.LinAlg.Adjoint{#s336,#s335} where #s335<:Base.LinAlg.AbstractTriangular where #s336, B::AbstractArray{T,2} where T) in Base.LinAlg at linalg/triangular.jl:1700)
```
requiring resolution signature
```julia
Tuple{typeof(*),Base.LinAlg.Adjoint{#s336,#s335} where #s335<:Base.LinAlg.AbstractTriangular where #s336,Base.LinAlg.Adjoint{#s336,#s335} where #s335<:(Union{Hermitian{T,S}, Hermitian{Complex{T},S}, Symmetric{T,S}} where S where T<:Real) where #s336}
```
But the resolution signatures for the 300 ambiguities with that commit get a bit abstract, e.g. ambiguous pair
```julia
(*(a::Adjoint, b::Adjoint) in Base.LinAlg at linalg/adjtrans.jl:105, *(adjA::Adjoint{#s336,#s335} where #s335<:Base.LinAlg.AbstractTriangular where #s336, B::AbstractArray{T,2} where T) in Base.LinAlg at linalg/triangular.jl:1700)
```
requiring resolution signature
```julia
Tuple{typeof(*),Adjoint{#s336,#s335} where #s335<:Base.LinAlg.AbstractTriangular where #s336,Adjoint}
```

I begin to lean towards simply not transferring those (overly broad?) fallbacks. Thoughts? Thanks!

**Comment by Sacha0** [comment]
Update, and a fundamental design point (opt-in versus opt-out laziness):

Transferring the broad fallbacks for `A[ct]_(mul|ldiv|rdiv)_B[ct][!]` from base/operators.jl to the corresponding `*`/`/`/`\`/`mul!`/`ldiv!`/`rdiv!` functions seems practically untenable due to the resulting ambiguities: Resolving one such ambiguity often introduces O(1)-O(10) more, so resolving both the initial 300-some and all downstream ambiguities likely would require substantially more than 300-some methods. Moreover, the resulting collection of methods would be brittle / difficult to work with at best.

So instead I added the shims for those broad fallbacks, e.g. `Ac_mul_B(a, b) = *(Adjoint(a), b)`, without transferring the associated methods to the corresponding functions. This changed dispatch, causing O(100) errors (mostly `MethodError`s, many differing only in operand eltypes). Fortunately those errors were relatively straightforward to fix, and the present version of this branch passes the linalg, sparse, and stdlib test suites locally. This leaves about 100 (relatively reasonable) ambiguities to fix.

What are the implications of not transferring those fallbacks?

Formerly, due to those broad fallbacks, most `A[ct]_(mul|ldiv|rdiv)_B[ct]` calls that did not directly hit appropriate `A[ct]_(mul|ldiv|rdiv)_B[ct]` specializations would eagerly adjoint/transpose the operands and then hit a decent specialization (instead of directly hitting a relatively slow generic `AbstractArray` method). The chief advantage of transferring those broad fallbacks was retaining that behavior. But there's another way to achieve similar behavior in practice:

This pull request presently makes lazy `Adjoint`/`Transpose` opt-out by type, in that if one does not want to define specializations over `Adjoint{...,MyType}`/`Transpose{...,MyType}`, one must define `Adjoint(A::MyType) = adjoint(A)` and likewise for `Transpose`. But laziness could instead be opt-in by type, and after some thought that might be much better. That is, rather than having the generic `Adjoint(A)` method construct an `Adjoint`-wrapped `A::MyType`, instead have the eager fallback `Adjoint(A) = adjoint(A)`, and if one wants to opt-in to laziness, one must define `Adjoint(A::MyType) = _Adjoint(A)` (where the generic `_Adjoint(...)` method creates an `Adjoint`-wrapped `A::MyType`) and specializations over `Adjoint{...,MyType}`. Then (1) behavior like the nice former `A[ct]_(mul|ldiv|rdiv)_B[ct]` fallback behavior described above happens mostly by default, (2) we achieve that behavior while avoiding the combinatorial ambiguity explosion that transferring the broad fallbacks involves, (3) the eager->lazy adjoint/transpose transition becomes less breaking overall and easier for users, and (4) after the transition, users need pay attention to lazy vs. eager adjoint/transpose primarily only for optimization (rather than of necessity).

Given those upsides and the problems associated with transferring the broad fallbacks, I presently lean heavily towards not transferring the broad fallbacks and making laziness opt-in. By default I will follow that approach in this and subsequent PRs and flip the opt-in switch at some point in that process. Best!

**Comment by fredrikekre** [comment]
> But laziness could instead be opt-in by type, and after some thought that might be much better. That is, rather than having the generic `Adjoint(A)` method construct an `Adjoint`-wrapped `A::MyType`, instead have the eager fallback `Adjoint(A) = adjoint(A)`, and if one wants to opt-in to laziness, one must define `Adjoint(A::MyType) = _Adjoint(A)` (where the generic `_Adjoint(...)` method creates an `Adjoint`-wrapped `A::MyType`) and specializations over `Adjoint{...,MyType}`.

I like the idea of lazy-opt in by type, but I think it is weird that `Adjoint(A)` (which essentially is a type constructor) does not *always* return an `Adjoint`. May I instead propose that one opts-in to lazyness via `adjoint`, such that you opt-in by defining `adjoint(::MyArray) = Adjoint(A)`?

**Comment by Sacha0** [comment]
Agreed, having `Adjoint`/`Transpose` behave strictly as constructors would be better. The idea above was primarily motivated by avoiding additional breakage. But this transition will be sufficiently breaking that perhaps avoiding additional breakage should be weighted less than achieving a better end result. I hope to post some evolved thoughts on the design space later today. But first...

The former version of this branch passed CI excepting the method ambiguity check. With benefit of ~75 disambiguating methods, the just-pushed version should also pass the method ambiguity check 🎉.

With the added disambiguating methods, adding the broad fallbacks for `A[ct]_(mul|ldiv|rdiv)_B[ct]` to `*`, `/`, and `\` superficially yields ~190 ambiguities (down from 300+). So progress independent of which route we ultimately go.

While continuing to think through the design space, I plan to complete and post as much work as possible that must happen in any case. Might be revealing. Best!

**Comment by StefanKarpinski** [comment]
@Sacha0, you are a saint.

**Comment by Sacha0** [comment]
Thank you for the kind words! Perhaps withhold judgement till this project comes together though... 😄

**Comment by Sacha0** [comment]
#### Update on the code in this branch:

Ground through the additional method ambiguitites that the recursive check revealed. Added commits replacing all `A[ct]_(mul|ldiv|rdiv)_B[ct][!]` definitions in `stdlib`. Consequently the present version of this branch passes the linalg, sparse, and stdlib test suites locally as well as the recursive method ambiguity check. So what say you now, CI?

#### Update on the broad fallbacks transfer issue:

With further investigation, transferring the broad `A[ct]_(mul|ldiv|rdiv)_B[ct]` fallbacks from base/operators.jl to `*`, `/`, and `\` turns out not only problematic but also mostly useless: More specific but still broad fallbacks for `*`, `/`, and `\` already exist (e.g. for `*(A::AbstractVecOrMat, B::AbstractVecOrMat)`), and `Adjoint`/`Tranpose` are `<:AbstractMatrix`, so fallbacks like `*(A::Adjoint, B)` essentially never get hit. (Concretely, the test suite hit the transferred fallbacks only ~three times.) So nevermind that approach, as transferring the broad fallbacks would not meaningfully preserve those fallbacks' advantageous behavior under the present system.

#### Observations for the future from the ambiguity elimination process:

- Broad fallbacks like `*(A::AbstractVecOrMat, B::AbstractVecOrMat)` generate most ambiguities. In the future, segregating such broad fallbacks in a separate name (or something similar) might be advantageous. For example, defining `*(a, b) = mulfallback(a, b)` (which would capture calls only if no other matching signature exists) and then [cascading](https://docs.julialang.org/en/stable/manual/methods/#Dispatch-on-one-argument-at-a-time-1) in `mulfallback` might improve things substantially. (Or perhaps cascading in `*` itself in `Base`.)

- Many method signatures are looser than they should be, though in such cases often necessarily given what the type system / dispatch can presently express. For example, `mul!(A::QRPackedQ, B::AbstractVecOrMat)` also implicitly requires that all of `B`'s entries are mutable, and somehow exposing that restriction to dispatch would help avoid many ambiguities.

- Many methods exist only to throw an error, and a decent fraction of those only to resolve ambiguities. Ironically, some such methods' signatures are quite broad, generating ambiguities in turn. Having some way to express "this signature should not exist, don't consider it in method resolution" or something similar to avoid this situation would be nice.

- Method ambiguities between two-arg `mul!` methods nicely highlight those methods' semantic ambiguity: Which argument does `mul!(A::Transpose{<:Any,<:Diagonal}, B::Transpose{<:Any,<:Diagonal})` mutate?

#### Observation for future refinement of lazy adjoint/transpose types:

- `Adjoint`/`Transpose` are `<:AbstractMatrix`, but when wrapping e.g. `Factorization`s or `Rotation`s this subtype relation is not correct. I am not certain how to avoid this issue without additional types and complexity or something along the lines of traits or abstract multiple inheritance.

#### Plan:

Move on to replacing `A[ct]_(mul|ldiv|rdiv)_B[ct][!]` calls (rather than defs) and at some point write up thoughts re. remaining design questions. Best!

**Comment by Sacha0** [comment]
So much green!

**Comment by andreasnoack** [comment]
First of all, this is awesome. Thank you so much for doing this work. It looks like you are close now. I'll look more carelly later this week. For now just a small comment

> Which argument does `mul!(A::Transpose{<:Any,<:Diagonal}, B::Transpose{<:Any,<:Diagonal})` mutate?

It should just throw. The method is not practically relevant.

Rename the function if think it helps anything. I don't. It is a function that multiply two matrices where one of the arguments is modified so I think the name `mul!` describes the purpose well but I don't really want to fight this fight anymore. Surprisingly many people that have newer used this function, and will newer use the function or encounter any problems from this function still have a lot of issues with the name of this function and we have spent way too much time discussing a function that only I and a few extert users find useful.

**Comment by Sacha0** [comment]
> It should just throw. The method is not practically relevant.

Agreed, and the disambiguation method is defined accordingly :). And yes, let's not re-litigate #16772 here. Best!

**Comment by Sacha0** [comment]
This pull request now replaces/rewrites all `A[ct]_(mul|ldiv|rdiv)_B[ct][!]` definitions and calls in `base/`, `test/`, and `stdlib/`, completing items 1-6 in the OP's updated task list.

`@andreasnoack` (thanks again!) and I discussed how to best move forward earlier today. The OP's task list has been updated accordingly, and tomorrow morning PT I plan to give this pull request another pass, make any necessary touchups, merge, and then approach tasks 7-10 in a followup pull request. Best!

File: base/linalg/adjtrans.jl [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+end
+struct Transpose{T,S} <: AbstractMatrix{T}
+    parent::S
+    function Transpose{T,S}(A::S) where {T,S}
+        checkeltype(Transpose, T, eltype(A))
+        new(A)
+    end
+end
+
+@pure function checkeltype(::Type{Transform}, ::Type{ResultEltype}, ::Type{ParentEltype}) where {Transform, ResultEltype, ParentEltype}
```
Comment (vchuravy) [inline base/linalg/adjtrans.jl]: Why does this need to be `@pure`

Comment (vtjnash) [inline base/linalg/adjtrans.jl]: This is very strongly not a `pure` function and terrible things will happen.

Comment (Sacha0) [reply base/linalg/adjtrans.jl]: This definition was preexisting for `RowVector`s :).

Comment (Sacha0) [reply base/linalg/adjtrans.jl]: As above, this definition was preexisting for `RowVector`s :).

**Comment by Jutho** [comment]
Thanks this is great. However, I thought this PR would come together with the parser changes that no longer parse `A'*B' as `Ac_mul_B`. Am I mistaken that it cannot be the plan that the following no longer works:
```julia
julia> struct MyType{T<:Number}
           a::T
       end

julia> Base.:*(x::MyType,y::MyType) = MyType(x.a*y.a)

julia> Base.adjoint(x::MyType) = x

julia> x=MyType(3)
MyType{Int64}(3)

julia> x'*x
ERROR: MethodError: no method matching *(::Base.LinAlg.Adjoint{Any,MyType{Int64}}, ::MyType{Int64})
Closest candidates are:
  *(::Any, ::Any, ::Any, ::Any...) at operators.jl:469
  *(::AbstractArray, ::Number) at arraymath.jl:55
  *(::AbstractArray{T,2} where T, ::Base.LinAlg.Transpose{#s344,#s343} where #s343<:RowVector where #s344) at linalg/rowvector.jl:221
  ...
Stacktrace:
 [1] Ac_mul_B(::MyType{Int64}, ::MyType{Int64}) at ./deprecated.jl:2968
 [2] top-level scope
```

Could we place have `adjoint` and `tranpose` in the fallback methods? Or are custom types supposed to overload `Adjoint` and `Transpose`? It feels strange to overload type constructors. 

Or is that what #25083 is about?

**Comment by Sacha0** [comment]
> Or is that what #25083 is about?

Yes, #25083 and downstream PRs continue this work :).

**Comment by Jutho** [comment]
Ok thanks. I don't see the parser changes there either yet, but I guess those are to come then? Thanks again for all the work!

**Comment by Sacha0** [comment]
Please see the OP for a task list. Best!

**Pull request merged**